// indexed/lib/methods.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.indexesMethodsKeys = exports.mutableMethodsKeys = exports.immutableMethodsKeys = exports.allMethods = exports.mutableMethods = exports.immutableMethods = exports.indexesMethods = undefined;

var _arrayNonMutative = require('./array-non-mutative');

var _arrayIndexes = require('./array-indexes');

var _arrayMutative = require('./array-mutative');

var indexesMethods = exports.indexesMethods = {
	reindex: _arrayIndexes.reindex,
	indexes: _arrayIndexes.indexes,
	addIndex: _arrayIndexes.addIndex
};

var immutableMethods = exports.immutableMethods = {
	some: _arrayNonMutative.some,
	map: _arrayNonMutative.map,
	fill: _arrayNonMutative.fill,
	find: _arrayNonMutative.find,
	get: _arrayNonMutative.get,
	getIndex: _arrayNonMutative.getIndex,
	has: _arrayNonMutative.has,

	findIndex: _arrayIndexes.findIndex,
	findIndexes: _arrayIndexes.findIndexes
};

var mutableMethods = exports.mutableMethods = {
	concat: _arrayMutative.concat,
	filter: _arrayMutative.filter,
	forEach: _arrayMutative.forEach,
	transform: _arrayMutative.transform,
	pop: _arrayMutative.pop,
	push: _arrayMutative.push,
	reverse: _arrayMutative.reverse,
	shift: _arrayMutative.shift,
	slice: _arrayMutative.slice,
	sort: _arrayMutative.sort,
	splice: _arrayMutative.splice,
	unshift: _arrayMutative.unshift,
	remove: _arrayMutative.remove,
	set: _arrayMutative.set,
	replace: _arrayMutative.replace,
	setMany: _arrayMutative.setMany,
	removeMany: _arrayMutative.removeMany,
	findMany: _arrayMutative.findMany,
	fromJson: _arrayMutative.fromJson,
	clear: _arrayMutative.clear
};

var allMethods = exports.allMethods = {
	some: _arrayNonMutative.some,
	map: _arrayNonMutative.map,
	fill: _arrayNonMutative.fill,
	find: _arrayNonMutative.find,
	get: _arrayNonMutative.get,
	getIndex: _arrayNonMutative.getIndex,
	has: _arrayNonMutative.has,

	indexes: _arrayIndexes.indexes,
	reindex: _arrayIndexes.reindex,
	findIndex: _arrayIndexes.findIndex,
	findIndexes: _arrayIndexes.findIndexes,
	addIndex: _arrayIndexes.addIndex,

	concat: _arrayMutative.concat,
	filter: _arrayMutative.filter,
	forEach: _arrayMutative.forEach,
	transform: _arrayMutative.transform,
	pop: _arrayMutative.pop,
	push: _arrayMutative.push,
	reverse: _arrayMutative.reverse,
	shift: _arrayMutative.shift,
	slice: _arrayMutative.slice,
	sort: _arrayMutative.sort,
	splice: _arrayMutative.splice,
	unshift: _arrayMutative.unshift,
	remove: _arrayMutative.remove,
	set: _arrayMutative.set,
	replace: _arrayMutative.replace,
	setMany: _arrayMutative.setMany,
	removeMany: _arrayMutative.removeMany,
	findMany: _arrayMutative.findMany,
	fromJson: _arrayMutative.fromJson
};

var immutableMethodsKeys = exports.immutableMethodsKeys = Object.keys(immutableMethods);
var mutableMethodsKeys = exports.mutableMethodsKeys = Object.keys(mutableMethods);
var indexesMethodsKeys = exports.indexesMethodsKeys = Object.keys(indexesMethods);
